File: vulnerable-code.js
// Arquivo JavaScript com vulnerabilidades para teste SAST

// 1. SQL Injection vulnerability
function getUserData(userId) {
  const query = "SELECT * FROM users WHERE id = " + userId; // Vulnerable to SQL injection
  return database.query(query);
}

// 2. XSS vulnerability
function renderUserName(userName) {
  document.getElementById("user").innerHTML = userName; // Vulnerable to XSS
}

// 3. Hardcoded secret (já detectado por outras ferramentas)
const API_KEY = "sk_live_1234567890abcdefghijk";

// 4. Command Injection vulnerability
const { exec } = require("child_process");
function processFile(filename) {
  exec("ls -la " + filename); // Vulnerable to command injection
}

// 5. Insecure random number generation
function generateToken() {
  return Math.random().toString(36); // Cryptographically insecure
}

// 6. Path Traversal vulnerability
const fs = require("fs");
function readFile(filepath) {
  return fs.readFileSync("./uploads/" + filepath); // Vulnerable to path traversal
}

// 7. Eval usage (dangerous)
function processUserInput(userCode) {
  return eval(userCode); // Never use eval with user input
}

// 8. Weak crypto
const crypto = require("crypto");
function hashPassword(password) {
  return crypto.createHash("md5").update(password).digest("hex"); // MD5 is weak
}

module.exports = {
  getUserData,
  renderUserName,
  processFile,
  generateToken,
  readFile,
  processUserInput,
  hashPassword,
};

File: vulnerable-code.py
# Arquivo Python com vulnerabilidades para teste SAST

import os
import subprocess
import hashlib
import pickle
import sqlite3
import random
from flask import Flask, request

app = Flask(__name__)

# 1. SQL Injection vulnerability
def get_user_data(user_id):
    conn = sqlite3.connect('users.db')
    cursor = conn.cursor()
    # Vulnerable to SQL injection
    query = f"SELECT * FROM users WHERE id = {user_id}"
    cursor.execute(query)
    return cursor.fetchall()

# 2. Command Injection vulnerability
def process_file(filename):
    # Vulnerable to command injection
    os.system(f"ls -la {filename}")

# 3. Hardcoded credentials
DATABASE_PASSWORD = "admin123"  # Hardcoded password
SECRET_KEY = "super-secret-key-123"

# 4. Insecure deserialization
def load_user_data(data):
    # Vulnerable to arbitrary code execution
    return pickle.loads(data)

# 5. Weak cryptographic hash
def hash_password(password):
    # MD5 is cryptographically weak
    return hashlib.md5(password.encode()).hexdigest()

# 6. Insecure random number generation
def generate_session_id():
    # Not cryptographically secure
    return str(random.randint(1000000, 9999999))

# 7. Path Traversal vulnerability
@app.route('/file')
def read_file():
    filename = request.args.get('file')
    # Vulnerable to path traversal
    with open(f"./uploads/{filename}", 'r') as f:
        return f.read()

# 8. Shell=True in subprocess (dangerous)
def execute_command(cmd):
    # Vulnerable to command injection
    subprocess.call(cmd, shell=True)

# 9. Assert used for authentication (bad practice)
def authenticate_user(username, password):
    # Assert can be disabled in production
    assert username == "admin" and password == "secret"
    return True

# 10. Eval usage (dangerous)
def calculate_expression(expression):
    # Never use eval with user input
    return eval(expression)

if __name__ == "__main__":
    # Debug mode should not be used in production
    app.run(debug=True, host='0.0.0.0')
